File: client/client-multi/client-objective-c/Kaa/transact/TransactionId.m
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "TransactionId.h"
#import "UUID.h"

@interface TransactionId ()

@property (nonatomic, strong) NSString *identifier;

@end

@implementation TransactionId

- (instancetype)init {
    return [self initWithStringId:[UUID randomUUID]];
}

- (instancetype)initWithStringId:(NSString *)stringId {
    self = [super init];
    if (self) {
        self.identifier = stringId;
    }
    return self;
}

- (instancetype)initWithTransactionId:(TransactionId *)transactionId {
    return [self initWithStringId:transactionId.identifier];
}

- (id)copyWithZone:(NSZone *)zone {
    return [[[self class] allocWithZone:zone] initWithStringId:self.identifier];
}

- (NSUInteger)hash {
    int prime = 31;
    return prime + [self.identifier hash];
}

- (BOOL)isEqual:(id)object {
    if ([object isKindOfClass:[TransactionId class]]) {
        TransactionId *other = (TransactionId *)object;
        if ([self.identifier isEqualToString:other.identifier]) {
            return YES;
        }
    }
    
    return NO;
}

- (NSString *)description {
    return self.identifier;
}

@end

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/transact/TransactionId.java
/*
 * Copyright 2014-2016 CyberVision, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.kaaproject.kaa.client.transact;

import java.util.UUID;

/**
 * Class representing unique transaction id for transactions initiated using
 * {@link Transactable}.<br>
 */
public final class TransactionId {
  private final String id;

  /**
   * Default constructor. Generates random unique transaction id.
   */
  public TransactionId() {
    this.id = UUID.randomUUID().toString();
  }

  /**
   * Copy constructor. Copies id from other TransactionId object.
   *
   * @param trxId the trx id
   */
  public TransactionId(TransactionId trxId) {
    this.id = trxId.id;
  }

  /**
   * Constructs object with predefined id.
   *
   * @param id the id
   */
  public TransactionId(String id) {
    this.id = id;
  }

  @Override
  public int hashCode() {
    final int prime = 31;
    int result = 1;
    result = prime * result + ((id == null) ? 0 : id.hashCode());
    return result;
  }

  @Override
  public boolean equals(Object obj) {
    if (this == obj) {
      return true;
    }
    if (obj == null) {
      return false;
    }
    if (getClass() != obj.getClass()) {
      return false;
    }
    TransactionId other = (TransactionId) obj;
    if (id == null) {
      if (other.id != null) {
        return false;
      }
    } else if (!id.equals(other.id)) {
      return false;
    }
    return true;
  }

  @Override
  public String toString() {
    return id;
  }

}
